programs/rewards/src/state/mining.rs

```diff
@@ -81,6 +81,35 @@ impl<'a> WrappedMining<'a> {
 
         Ok(())
     }
+
+    /// Decrease rewards
+    pub fn decrease_rewards(&mut self, mut decreased_weighted_stake_number: u64) -> ProgramResult {
+        if decreased_weighted_stake_number == 0 {
+            return Ok(());
+        }
+
+        if decreased_weighted_stake_number > self.mining.share {
+            return Err(MplxRewardsError::DecreaseRewardsTooBig.into());
+        }
+
+        self.mining.share = self
+            .mining
+            .share
+            .safe_sub(decreased_weighted_stake_number)?;
+
+        for (_, stake_diff) in self.weighted_stake_diffs.iter_mut().rev() {
```

Comment:
A little comment on what's going on here is much appreciated. As far as I got it - we go backwards through the rbtree and update the last diff to apply penalty.